mm2src/coins/eth/eth_balance_events.rs

```diff
@@ -1,28 +1,58 @@
+use super::EthCoin;
+use crate::{eth::{u256_to_big_decimal, Erc20TokenInfo},
+            BalanceError, CoinWithDerivationMethod};
+use common::{executor::Timer, log, Future01CompatExt};
+use mm2_err_handle::prelude::MmError;
+use mm2_event_stream::{Broadcaster, Event, EventStreamer, NoDataIn, StreamHandlerInput};
+use mm2_number::BigDecimal;
+
 use async_trait::async_trait;
-use common::{executor::{AbortSettings, SpawnAbortable, Timer},
-             log, Future01CompatExt};
 use ethereum_types::Address;
-use futures::{channel::oneshot::{self, Receiver, Sender},
-              stream::FuturesUnordered,
-              StreamExt};
+use futures::{channel::oneshot, stream::FuturesUnordered, StreamExt};
 use instant::Instant;
-use mm2_core::mm_ctx::MmArc;
-use mm2_err_handle::prelude::MmError;
-use mm2_event_stream::{behaviour::{EventBehaviour, EventInitStatus},
-                       ErrorEventName, Event, EventName, EventStreamConfiguration};
-use mm2_number::BigDecimal;
+use serde::Deserialize;
+use serde_json::Value as Json;
 use std::collections::{HashMap, HashSet};
 
-use super::EthCoin;
-use crate::{eth::{u256_to_big_decimal, Erc20TokenInfo},
-            BalanceError, CoinWithDerivationMethod, MmCoin};
+#[derive(Deserialize)]
+#[serde(deny_unknown_fields, default)]
+struct EthBalanceStreamingConfig {
+    /// The time in seconds to wait before re-polling the balance and streaming.
+    pub stream_interval_seconds: f64,
+}
+
+impl Default for EthBalanceStreamingConfig {
+    fn default() -> Self {
+        Self {
+            stream_interval_seconds: 10.0,
+        }
+    }
+}
+
+pub struct EthBalanceEventStreamer {
+    /// The period in seconds between each balance check.
+    interval: f64,
+    coin: EthCoin,
+}
```

Comment:
because there could be many other streamers that would use `EthCoin` not just balance streaming, having the coin stored inside gives a consistent interface.

Reply:
> because there could be many other streamers that would use EthCoin not just balance streaming

Exactly. So if I have streamers like `Streamer1`, `Streamer2`, `Streamer3` .. n, why should I clone `EthCoin` into each one of them instead of keeping them inside some field in `EthCoin`?

Reply:
> Exactly. So if I have streamers like Streamer1, Streamer2, Streamer3 .. n, why should I clone EthCoin into each one of them instead of keeping them inside some field in EthCoin?

u don't want to store tons of streamers inside each coin, this is just making it harder to maintain.
right now they self-own themselves and any object they need to operate successfully, e.g. `NetworkEvent` has `MmArc`.

Reply:
> right now they self-own themselves

Could you clarify why its preferable to keep streamer structs as self-own? Im not quite familiar with event streaming feature, so im studying the code now.

Reply:
streamers are autonomous background threads. we can have them work by themselves without the need to store them somewhere. this reduces the clutter over the code base as then we don't have to think about the storage of these streamers, they are kinda stored in the streaming manager in the sense that they could be terminated from there (via the shutdown handle).

one could in the simplest terms create a streamer with just the streamer impl + enabler endpoint without any change to the coin/object the streamer revolves around (that's the case for a simple periodic streamer), and some notification code related to the object it streams for if the streamer was reactive.

we also get the added point of easy self-cleanup when it dies.

Reply:
Honestly I strongly disagree on this. You don't have to write every streaming-channel related logic inside the coin modules. My point is it just looks very annoying to own coins inside the streaming types while they are entirely depended on the life of the cloned coin. You can keep it however you like, maybe it's just me.

Reply:
IMHO this idea of implementing streamer inside the object itself is more applicable to streamers like SwapStatusStreamer or RpcTaskManager in which we have is a flow of events so notify the subscriber at the event source.
For the other hand, for coin balance we need to do polling (no events to capture inside the coin code) and it's simpler to do it outside so we could use same polling code for different coins